File: workspace/scripts/cb_search.py
#!/usr/bin/env python3
"""
CodeBuddy Web Search Skill — Streamable HTTP + ACP protocol
"""
import os, sys, json, time, uuid, requests

API_KEY = "ck_fn3pyichksu8.pl8i-Kss-DRFCc496IJFG34Y7V5ea2o9DIQsLVKAg6U"
BASE_URL = "https://www.codebuddy.cn/v2/agentos"

def log(*args):
    print(*args, file=sys.stderr, flush=True)

def headless_main(query):
    # 1. Create Runtime
    manifest = {
        "id": f"ws-{uuid.uuid4().hex[:6]}",
        "name": "搜索", "manifestVersion": "1.0.0",
        "system_prompt": "你是搜索助手，用 web_search 搜索信息，用中文回答。",
        "skills": [{"name": "web_search"}],
        "secrets": [{"key": "CODEBUDDY_API_KEY", "value": API_KEY}]
    }
    
    h = {"Content-Type": "application/json", "X-Api-Key": API_KEY, "X-Source-App": "cloud-agent"}
    resp = requests.post(f"{BASE_URL}/runtimes", json={
        "runtimeName": f"搜:{query[:15]}", "agentManifest": manifest, "sandboxType": "CS",
    }, headers=h, timeout=30)
    rid = resp.json().get("data", {}).get("id")
    log(f"[Runtime] {rid}")
    
    # 2. Poll for ACP
    acp_url = token = None
    for i in range(15):
        time.sleep(5)
        r = requests.get(f"{BASE_URL}/runtimes/{rid}", headers=h, timeout=10)
        if r.status_code != 200: continue
        d = r.json().get("data", {})
        l = d.get("links", {}).get("acpLink", {})
        if l.get("url"):
            acp_url, token = l["url"], l["token"]
            break
        log(f"  wait ({i+1}/15)")
    if not acp_url:
        return {"error": "沙箱超时"}
    log(f"[ACP] 就绪")
    
    # 3. ACP: Streamable HTTP 协议
    # Step 1: POST /acp with initialize (POST creates the connection)
    # Ref: https://spec.modelcontextprotocol.io/specification/basic/
    # Actually ACP is similar to MCP's Streamable HTTP
    
    # Send initialize via POST
    init_body = json.dumps({
        "jsonrpc": "2.0", "id": 1,
        "method": "initialize",
        "params": {
            "protocolVersion": "0.1.0",
            "capabilities": {},
            "clientInfo": {"name": "openclaw-search", "version": "1.0.0"}
        }
    })
    
    # POST with both Accept headers
    resp = requests.post(acp_url, data=init_body, headers={
        "Content-Type": "application/json",
        "Authorization": f"Bearer {token}",
        "Accept": "application/json, text/event-stream",
    }, timeout=30)
    
    log(f"[ACP] Init: {resp.status_code}")
    if resp.status_code != 200:
        return {"error": f"ACP初始化失败: {resp.status_code} {resp.text[:200]}"}
    
    init_result = resp.json()
    log(f"[ACP] Init result: {json.dumps(init_result, ensure_ascii=False)[:200]}")
    
    if "error" in init_result:
        # Maybe need SSE connection first
        log("[ACP] 初始化失败，尝试SSE先连...")
        
        # Step 1b: GET SSE to establish connectionId
        sse = requests.get(acp_url, stream=True, headers={
            "Authorization": f"Bearer {token}",
            "Accept": "text/event-stream",
        }, timeout=30)
        
        conn_id = None
        for line in sse.iter_lines(decode_unicode=True):
            if line and line.startswith("data: "):
                try:
                    d = json.loads(line[6:])
                    conn_id = d.get("connectionId", d.get("data", {}).get("connectionId"))
                except: pass
            if conn_id: break
        
        log(f"[ACP] ConnectionId: {conn_id}")
        
        # Now send POST with session/new
        sn = json.dumps({
            "jsonrpc": "2.0", "id": 2,
            "method": "session/new",
            "params": {"cwd": "/workspace", "mcpServers": []}
        })
        resp2 = requests.post(acp_url, data=sn, stream=True, timeout=120, headers={
            "Content-Type": "application/json",
            "Authorization": f"Bearer {token}",
            "Accept": "application/json, text/event-stream",
            "Acp-Connection-Id": conn_id,
        })
        log(f"[ACP] session/new: {resp2.status_code}")
        result2 = resp2.json()
        log(f"[ACP] sess result: {json.dumps(result2, ensure_ascii=False)[:200]}")
        
        sid = result2.get("sessionId") or result2.get("result", {}).get("sessionId")
        if not sid:
            return {"error": f"Session创建失败", "detail": str(result2)[:300]}
        
        # Prompt
        prompt = json.dumps({
            "jsonrpc": "2.0", "id": 3,
            "method": "session/prompt",
            "params": {
                "session_id": sid,
                "prompt": [{"role": "user", "content": [{"type": "text", "text": query}]}],
            }
        })
        resp3 = requests.post(acp_url, data=prompt, stream=True, timeout=300, headers={
            "Content-Type": "application/json",
            "Authorization": f"Bearer {token}",
            "Accept": "application/json, text/event-stream",
            "Acp-Connection-Id": conn_id,
        })
        
        log(f"[ACP] Prompt: {resp3.status_code}")
        result_text = ""
        for line in resp3.iter_lines(decode_unicode=True):
            if not line: continue
            if line.startswith("data: "):
                try:
                    d = json.loads(line[6:])
                    if "error" in d:
                        log(f"[ACP] Error: {d['error']}")
                        break
                    # Chunks come as notifications
                except: pass
            elif line.startswith("event: "):
                pass
        
        result_text = resp3.text if not result_text else result_text
        if not result_text:
            try:
                result_text = json.dumps(resp3.json(), ensure_ascii=False)
            except:
                result_text = resp3.text[:5000]
        
        sse.close()
        return {"query": query, "result": result_text[:10000], "source": "codebuddy"}
    
    # If init worked directly, proceed...
    sid = init_result.get("sessionId")
    if not sid:
        # Normal flow: send session/new
        pass
    
    return {"error": "需要更多调试"}

if __name__ == "__main__":
    q = " ".join(sys.argv[1:]) or "2025年全球人工智能市场规模"
    from concurrent.futures import ThreadPoolExecutor
    result = headless_main(q)
    print(json.dumps(result, ensure_ascii=False, indent=2))

File: workspace/scripts/cb_search.mjs
#!/usr/bin/env node
/**
 * CodeBuddy Web Search Agent — SDK AcpClient 直连
 * 
 * 绕过 Session REST API，直接用 ACP 的 session/new 创建会话
 */
import { CloudAgentClient, ManifestBuilder } from '@tencent-ai/cloud-agent-sdk';
const API_KEY = 'ck_fn3pyichksu8.pl8i-Kss-DRFCc496IJFG34Y7V5ea2o9DIQsLVKAg6U';

async function search(query, timeoutMs = 180000) {
  const client = new CloudAgentClient({ apiKey: API_KEY, logLevel: 'warn', timeoutMs });
  let runtime;

  // 1. Create Runtime
  const manifest = new ManifestBuilder()
    .id(`cb-${Date.now().toString(36)}`).name('S').version('1.0.0')
    .systemPrompt('你是搜索助手，用 web_search 搜索信息，用中文回答并注明来源。')
    .skills('web_search').build();

  runtime = await client.runtimes.create({
    runtimeName: `搜:${query.slice(0,12)}`,
    agentManifest: manifest,
    sandboxType: 'CS',
  });

  const rid = runtime.id;
  for (let i = 0; i < 12; i++) {
    await new Promise(r => setTimeout(r, 5000));
    runtime = await client.runtimes.get(rid);
    if (runtime._acpUrl) break;
  }
  if (!runtime._acpUrl) throw new Error('沙箱启动超时');

  // 2. 创建 Session（通过 REST），然后获取 _acpClient
  const session = await runtime.sessions.create({ sessionName: 's' });
  
  // 先 connect（无论是否成功，_acpClient 会创建）
  try { await Promise.race([session.connect(), new Promise((_,r)=>setTimeout(()=>r('ok'), 10000))]); } catch(e) {}

  if (!session._acpClient) throw new Error('ACP连接失败');

  // 3. 用 ACP 创建新 Session（绕过 loadSession）
  const newSid = await session._acpClient.sessionNew();
  if (!newSid) throw new Error('session/new 失败');
  Object.defineProperty(session, 'id', { value: newSid, writable: true });

  // 4. Prompt + 流式读取
  return new Promise((resolve, reject) => {
    const timeout = setTimeout(() => { session?.disconnect(); reject(new Error('搜索超时')); }, timeoutMs);
    let result = '';

    session.prompt({
      text: query,
      onChunk: (chunk) => { process.stdout.write(chunk?.text || ''); result += chunk?.text || ''; },
    }).then(resp => {
      clearTimeout(timeout);
      session?.disconnect();
      resolve({ result, stopReason: resp.stopReason });
    }).catch(err => {
      clearTimeout(timeout);
      session?.disconnect();
      reject(err);
    });
  });
}

const q = process.argv.slice(2).join(' ') || '搜索2025年全球人工智能市场规模数据';
search(q).then(r => console.log('\n' + JSON.stringify({ query: q, result: r.result, source: 'codebuddy' })))
  .catch(e => { console.error('\nError:', e.message); console.log(JSON.stringify({ error: e.message })); process.exit(1); });
